### ISSUE TITLE
Action Route working well in development, but not being recognized in Production

### ISSUE THREAD

**Issue by HadiMalikDev** [Shopify/shopify-app-template-remix#559]
# Issue summary

I'm trying to call an action function from one of my route files. The entire thing works well in development, but the API call yields a 404 on production on the shopify app. Even more puzzling is the fact that if I try to call the API from an external source(ReqBin), the call succeeds. This makes me think that the issue is originating from the App bridge
![image](https://github.com/Shopify/shopify-app-template-remix/assets/78081816/4ac19cd9-9143-4902-aa0d-134b8fb672a8)

The first POST call was one made from my app, which yielded a 404.
The second POST call, that yields a 204, was made from ReqBin. 

Attached are the error logs I see in the browser 
![image](https://github.com/Shopify/shopify-app-template-remix/assets/78081816/2580247e-70fe-429c-8b85-2afc6ea6aa51)

- `@shopify/shopify-app-remix` version: 1.2.1
- Node version: 18.13.0
- Operating system: MacOS Monterey

If it helps, the issue cropped up after I suddenly received an error regarding the `isbot` package in the `entry.server.jsx`, which suddenly did not have a default export and had to be imported via `import {isbot} from "isbot"` rather than the default `import isbot from "isbot"`

## Expected behavior
The API call should work both from within the app, and externally, both in development and production
## Actual behavior
The API call yields a 404 when called within the app during production, but works normally when called externally

**Comment by paulomarg** [comment]
This is a strange one! I don't believe we throw 404s from the remix package, so it's hard to say what might be going wrong here. If it is somehow related to the package, it would probably be helpful to turn on debug logging, if that's possible.

You can go into `shopify.server` and add this to your config:

```ts
import {LogSeverity} from '@shopify/shopify-api';

shopifyApp({
  logger: {level: LogSeverity.Debug},
})
```

If this is a problem with the package / app template we can help fix it.

**Comment by HadiMalikDev** [comment]
Hey @paulomarg . Thank you for the response. I managed to pinpoint the issue. It's not an issue with the Remix package, but with the Shopify API, I believe. The Theme API works perfectly locally, but throws a 404 on a `PUT` request on production, which gets propagated up. I've asked for, and gotten an exemption from Shopify to use the Theme API, so I'm confused as to what could be causing the issue.  I double-checked the SHOPIFY_API_KEY and SHOPIFY_API_SECRET and the ones I set in production match the ones I set locally. Any ideas?

**Comment by paulomarg** [comment]
I don't believe we throw 404s out of `shopify-api` either, but if it's happening during the admin authentication step (which _might_ be the case based on the logs), it'll be hard to tell where it goes wrong without debug logs.

**Comment by HadiMalikDev** [comment]
Hey @paulomarg !

Finally got around to looking at it

`
2024-03-20T09:58:26.269822+00:00 app[web.1]: [shopify-app/DEBUG] Validating session token
2024-03-20T09:58:26.271229+00:00 app[web.1]: [shopify-app/DEBUG] Session token is valid | {payload: {"iss":"https://themeable-labs.myshopify.com/admin","dest":"https://themeable-labs.myshopify.com","aud":"13f583ac692336828bd33d1c6d552e04","sub":"103281688891","exp":1710928741,"nbf":1710928681,"iat":1710928681,"jti":"d015a4db-0cb8-4168-b305-257bb248847e","sid":"d5525230-189b-46ca-95de-2ceffc72c0c2","sig":"165a05667883d80ac675c881fd50f2ed2d4b7f86662100a683200932f7b5033b"}}
2024-03-20T09:58:26.271293+00:00 app[web.1]: [shopify-app/DEBUG] Session token is valid | {shop: themeable-labs.myshopify.com, payload: [object Object]}
2024-03-20T09:58:26.271351+00:00 app[web.1]: [shopify-app/DEBUG] Loading session from storage | {sessionId: offline_themeable-labs.myshopify.com}
2024-03-20T09:58:26.275046+00:00 app[web.1]: [shopify-app/DEBUG] Found a valid session | {shop: themeable-labs.myshopify.com}
2024-03-20T09:58:26.275102+00:00 app[web.1]: [shopify-app/DEBUG] Request is valid, loaded session from session token | {shop: themeable-labs.myshopify.com, isOnline: false}
2024-03-20T09:58:26.275500+00:00 app[web.1]: AFTER AUTHENTICATION
2024-03-20T09:58:26.279541+00:00 app[web.1]: SECTION {
2024-03-20T09:58:26.279542+00:00 app[web.1]:   id: '0281816c-4ce9-4a66-8d35-f46a40f0a11b',
2024-03-20T09:58:26.279543+00:00 app[web.1]:   name: 'Section 02',
2024-03-20T09:58:26.279543+00:00 app[web.1]:   desc: 'Very good section, vry amazing grape',
2024-03-20T09:58:26.279543+00:00 app[web.1]:   handle: 'section-02',
2024-03-20T09:58:26.279544+00:00 app[web.1]:   preview_image: 'https://picsum.photos/200',
2024-03-20T09:58:26.279544+00:00 app[web.1]:   main_image: 'https://picsum.photos/1000/300',
2024-03-20T09:58:26.279544+00:00 app[web.1]:   file_name: 'section-02.liquid',
2024-03-20T09:58:26.279546+00:00 app[web.1]:   file_url: 'https://firebasestorage.googleapis.com/v0/b/themeable-files.appspot.com/o/section-02.liquid?alt=media&token=d30bd3d9-7f2f-40cc-b931-902feacc52f4'
2024-03-20T09:58:26.279547+00:00 app[web.1]: }
2024-03-20T09:58:26.279585+00:00 app[web.1]: METHOD POST
2024-03-20T09:58:26.279659+00:00 app[web.1]: BODY { theme: 158373544251, section: 'section-02' }
2024-03-20T09:58:26.279894+00:00 app[web.1]: ASSET CREATED
2024-03-20T09:58:26.280639+00:00 app[web.1]: [shopify-api/DEBUG] REST client has a redundant API version override to the default 2024-01
2024-03-20T09:58:26.434504+00:00 app[web.1]: [shopify-app/DEBUG] Got an HTTP response error from the API: Received an error response (404 Not Found) from Shopify:
2024-03-20T09:58:26.434506+00:00 app[web.1]: "Not Found"
2024-03-20T09:58:26.434508+00:00 app[web.1]: If you report this error, please include this id: 76f4ecb3-48e2-4301-b0ad-9f67bbdcf1e8-1710928706 | {code: 404, statusText: Not Found, body: {"errors":"Not Found"}}
2024-03-20T09:58:26.436253+00:00 app[web.1]: Response {
2024-03-20T09:58:26.436254+00:00 app[web.1]:   size: 0,
2024-03-20T09:58:26.436255+00:00 app[web.1]:   [Symbol(Body internals)]: {
2024-03-20T09:58:26.436256+00:00 app[web.1]:     body: ReadableStream {
2024-03-20T09:58:26.436256+00:00 app[web.1]:       _state: 'readable',
2024-03-20T09:58:26.436256+00:00 app[web.1]:       _reader: undefined,
2024-03-20T09:58:26.436258+00:00 app[web.1]:       _storedError: undefined,
2024-03-20T09:58:26.436259+00:00 app[web.1]:       _disturbed: false,
2024-03-20T09:58:26.436259+00:00 app[web.1]:       _readableStreamController: [ReadableStreamDefaultController]
2024-03-20T09:58:26.436259+00:00 app[web.1]:     },
2024-03-20T09:58:26.436260+00:00 app[web.1]:     type: 'text/plain;charset=UTF-8',
2024-03-20T09:58:26.436260+00:00 app[web.1]:     size: 22,
2024-03-20T09:58:26.436261+00:00 app[web.1]:     boundary: null,
2024-03-20T09:58:26.436261+00:00 app[web.1]:     disturbed: false,
2024-03-20T09:58:26.436262+00:00 app[web.1]:     error: null
2024-03-20T09:58:26.436262+00:00 app[web.1]:   },
2024-03-20T09:58:26.436263+00:00 app[web.1]:   [Symbol(Response internals)]: {
2024-03-20T09:58:26.436263+00:00 app[web.1]:     url: undefined,
2024-03-20T09:58:26.436263+00:00 app[web.1]:     status: 404,
2024-03-20T09:58:26.436263+00:00 app[web.1]:     statusText: '',
2024-03-20T09:58:26.436264+00:00 app[web.1]:     headers: { 'content-type': 'application/json; charset=utf-8' },
2024-03-20T09:58:26.436264+00:00 app[web.1]:     counter: 0,
2024-03-20T09:58:26.436264+00:00 app[web.1]:     highWaterMark: undefined
2024-03-20T09:58:26.436264+00:00 app[web.1]:   }
2024-03-20T09:58:26.436265+00:00 app[web.1]: }
2024-03-20T09:58:26.436434+00:00 app[web.1]: { errors: 'Not Found' }
2024-03-20T09:58:26.437949+00:00 app[web.1]: POST /app?embedded=1&hmac=98cf03c6e238ccc931051e6c9e5e144404c754931b18d364b34a7a0443e10584&host=YWRtaW4uc2hvcGlmeS5jb20vc3RvcmUvdGhlbWVhYmxlLWxhYnM&id_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJodHRwczpcL1wvdGhlbWVhYmxlLWxhYnMubXlzaG9waWZ5LmNvbVwvYWRtaW4iLCJkZXN0IjoiaHR0cHM6XC9cL3RoZW1lYWJsZS1sYWJzLm15c2hvcGlmeS5jb20iLCJhdWQiOiIxM2Y1ODNhYzY5MjMzNjgyOGJkMzNkMWM2ZDU1MmUwNCIsInN1YiI6IjEwMzI4MTY4ODg5MSIsImV4cCI6MTcxMDkyODc0MCwibmJmIjoxNzEwOTI4NjgwLCJpYXQiOjE3MTA5Mjg2ODAsImp0aSI6ImU2NzE4ZGU4LTYyMWEtNGE5MS04MDQ1LTYwZmFkYWM1NTcxYyIsInNpZCI6ImQ1NTI1MjMwLTE4OWItNDZjYS05NWRlLTJjZWZmYzcyYzBjMiIsInNpZyI6IjM5NTU0ZWE3YmZlMWZmYWQzZTU2NzEyNDU5NGZiNDM0NGE5ZTRjZGIwZDIwNzY2NjYxNjZjMGE5ZDBhZDM5OWQifQ.RuP-9qq5iGsqJbhAdB9nYktIriNoDuImB1yQIh8Ewws&locale=en-PK&session=e3bd8ffbbb8e1f5d34d805462b17e17a8ca42440f5ac97c373901dc113c38a35&shop=themeable-labs.myshopify.com&timestamp=1710928680&_data=routes%2Fapp 500 - - 169.975 ms
`

Here's a screenshot, too

<img width="398" alt="image" src="https://github.com/Shopify/shopify-app-template-remix/assets/78081816/16441afb-7f1a-4420-87b3-f976a3f730a5">

Thank you in advance!

**Comment by paulomarg** [comment]
Hey @HadiMalikDev, I touched base with the team and this is expected - you're lacking the `write_themes_assets` scope, which needs to be explicitly requested and approved by Shopify.

In order to be able to alter assets, you'll need to request an exemption, as outlined [in the docs](https://shopify.dev/docs/apps/online-store/other-integration-methods/asset#uses-cases-eligible-for-exemption).

Since there is no action to be taken in the template / packages, I'm going to close this issue. If you still have problems accessing this, I'd recommend going through Partner support because they'll be able to help more than I will.

Thanks for your patience in sorting this out, hope this helps!

**Comment by HadiMalikDev** [comment]
Hey @paulomarg !

Thank you so much for your help in all this. I figured out the issue. I already had the write_themes scope, and already had it approved by Shopify. The reason it wasn't working on production for me was because the production database was still using the tokens that were generated **before** I had gotten the exemption from Shopify. So, I'm assuming that the tokens generated by Shopify on auth also include if you have access to the write_themes API. It was working on local development since I had formatted the entire database after getting permission from Shopify, so that was why it wasn't being replicated in the local environment.

Nuking the production session table and having them re-generated on auth fixed the issue. Putting it here so others can have a look at it too. Thank you for all your help!
